homeassistant/components/sensor/imap.py

```diff
@@ -54,45 +63,107 @@ def __init__(self, name, user, password, server, port):
-            _LOGGER.info("Connection to %s lost, attempting to reconnect",
-                         self._server)
-            try:
-                self.connection = self._login()
-            except imaplib.IMAP4.error:
-                _LOGGER.error("Failed to reconnect.")
+    @property
+    def available(self):
+        """Return the availability of the device."""
+        return self._connection is not None
 
     @property
-    def icon(self):
-        """Return the icon to use in the frontend."""
-        return ICON
+    def should_poll(self):
+        """Return if polling is needed."""
+        return self._connection and not self._connection.has_capability('IDLE')
+
+    @asyncio.coroutine
+    def connection(self):
+        """Return a connection to the server, establishing it if necessary."""
+        import aioimaplib
+
+        if self._connection is None:
+            try:
+                self._connection = aioimaplib.IMAP4_SSL(
+                    self._server, self._port)
+                yield from self._connection.wait_hello_from_server()
+                yield from self._connection.login(self._user, self._password)
+                yield from self._connection.select()
+            except (aioimaplib.AioImapException, asyncio.TimeoutError):
+                self._connection = None
+
+        return self._connection
+
+    @asyncio.coroutine
+    def idle_loop(self):
+        """Wait for data pushed from server."""
+        import aioimaplib
+
+        while True:
+            try:
+                if (yield from self.connection()):
+                    yield from self.refresh_unread_count()
+                    yield from self.async_update_ha_state()
+
+                    idle = yield from self._connection.idle_start()
+                    yield from self._connection.wait_server_push()
+                    self._connection.idle_done()
+                    yield from asyncio.wait_for(idle, 10)
+                else:
+                    yield from self.async_update_ha_state()
+            except (aioimaplib.AioImapException, asyncio.TimeoutError):
+                self.disconnect()
+
+    @asyncio.coroutine
+    def async_update(self):
+        """Periodic polling of state."""
+        import aioimaplib
+
+        try:
+            if (yield from self.connection()):
+                yield from self.refresh_unread_count()
+        except (aioimaplib.AioImapException, asyncio.TimeoutError):
+            self.disconnect()
+
+    @asyncio.coroutine
+    def refresh_unread_count(self):
+        """Check the number of unread emails."""
+        if self._connection:
+            yield from self._connection.noop()
+            _, lines = yield from self._connection.search('UnSeen UnDeleted')
+            self._unread_count = len(lines[0].split())
+
+    def disconnect(self):
+        """Drop the connection after exceptions."""
+        _LOGGER.warning("Lost %s, attempting to reconnect", self._server)
```

Comment:
Strange warning to be inside a method called `disconnect`. Also, not much disconnecting going on here?

Reply:
I see your point but I'm not completely sure how to address it. I have renamed the method to `disconnected` since it cleans up after we realize that we cannot talk to the server any more. To clean up properly we should actually call `logout` (like in `shutdown`) but that is not possible when the connection is lost.

The warning does say that the connection was lost, "attempting to reconnect" is meant to make that sound less fatal.